rfcs/47-better-error-handing.md

```diff
@@ -0,0 +1,287 @@
+     - `NewSomeError(contextA, contextB)` where 
+       - `SomeError` is returned and it is defined as:
+			```go
+			type SomeError struct {
+				ContextA string
+				ContextB structB
+				...
+			}
+			```
+       - `SomeError` SHOULD implement `Error`
+       - `SomeError` CAN implement `Unwrap`, returning the category (or the label) of the error, which is a sentinel error
+     - `fmt.Errorf("%w: %v", SomeError, err)` if the expected error is caused by another `error` value
+       - `SomeError` is one of the two kind of errors above
+       - `err` is the original error
+
+Real examples of expected errors:
+
+Definitions:
+```go
+var (
+	// ErrCapabilityInsufficient means this service doesn't have this capability
+	ErrCapabilityInsufficient = errors.New("capability insufficient")
+
+	// ErrObjectNotExist means the object to be operated is not exist.
+	ErrObjectNotExist = errors.New("object not exist")
+)
+
+// NewPairUnsupportedError will create a new PairUnsupportedError.
+func NewPairUnsupportedError(pair types.Pair) *PairUnsupportedError {
+	return &PairUnsupportedError{
+		Pair: pair,
+	}
+}
+
+// PairUnsupportedError means this operation has unsupported pair.
+type PairUnsupportedError struct {
+	Pair types.Pair
+}
+
+func (e *PairUnsupportedError) Error() string {
+	return fmt.Sprintf("pair unsupported, %s: %s", e.Pair, ErrCapabilityInsufficient.Error())
+}
+
+// Unwrap implements xerrors.Wrapper
+func (e *PairUnsupportedError) Unwrap() error {
+	return ErrCapabilityInsufficient
+}
+```
+
+![](./47/new.png)
+
+In short, compared with current practice, there are three changes:
+1. Ad-hoc string errors SHOULD not be used.
+2. SDK errors SHOULD not be wrapped.
+3. Wrapped error `struct`s won't wrap another layer of error `struct`s, and its wrapped `Err` is hidden.
+
+### Where to define expected errors
+
+Top-level errors SHOULD be defined in [go-storage].
+
+For wrapped errors, if an error is related to a feature supported by service pairs, e.g., server-side encryption, it is considered a service-specific error and should be defined in `go-service-*`. Otherwise, it is considered a global error and should be defined in [go-storage].
+
+New global errors except the ones listed in this RFC should be defined via the RFC procedure.
+
+### Naming convention
+
+Except that
+- Top-level errors are named as `<noun>Error`, `StorageError`, `ServiceError`, `InitError`
+- Unexpected error is named as `ErrUnexpected`
+
+Error `struct`s SHOULD be named as `<noun><adj>Error`, and sentinel errors SHOULD be named as `Err<noun><adj>`.
+
+Because we can organize them via `<noun>` easily:
+- `ObjetNotExist`
+- `ObjectModeInvalid`
+
+instead of
+- `NotExistObject`
+- `InvalidObjectMode`
+
+### All currently available global expected errors 
+
+- ErrCapabilityInsufficient
+  - MetadataUnrecognizedError
+  - PairUnsupportedError
+- ErrRestrictionDissatisfied
+  - PairRequiredError
+- ErrObjectNotExist
+- ErrObjectModeInvalid
```

Comment:
Do we need to introduce a new error struct that specifies `expected mode` and `actual mode`?

Reply:
If an operation needs a specific mode, then `expected mode` is fine.
If an operation allows a set of different modes, how to handle this case?

Reply:
Object mode's design makes sure that object mode is orthogonal, so an operation that requires object mode check only allows a single mode here.

- WriteAppend requires `IsAppend`
- ListPart requires `IsPart`
- Read requires `IsRead` (not implemented, added in https://github.com/aos-dev/go-storage/issues/557)
- Write doesn't need object mode check

Reply:
What about this:
https://github.com/aos-dev/go-service-s3/blob/84e8d53888817a5508aaf54f260a8e9d1a7fb37b/storage.go#L173

Reply:
`ListMode` is also orthogonal, but I think return `actual mode` is enough for `ListModeInvalid` error.